Added option to sort project papers

Merseleo opened the pull request:
Closes #39 

## What I have made

<!-- write down what changes have been made, what was removed/added/changed -->
<!-- e.g. I added a new button, which does x -->

- I added a new button on the project papers overview page so the user can select the sorting option
- Accordingly to the selected sort option, the paper are sorted within the stages; currently the sorting options "Paper Id", "Paper title", "Year" and "Decision" are supported
- When the sort options changes, the new sort option is persisted in the URL and loaded from the URL as well

In general, most of this functionality can be applied to other pages or lists as well, as long as they deal with (project papers).

## Checklist

Either tick or cross out the items that do not apply (using \~\~example text\~\~) and give a reason why the item does not apply.

- [x] I have updated the documentation accordingly and commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
  - [x] unit tests
  - [x] integration tests
  - [x] end-to-end tests
- [ ] (for reviewer) I have checked the implementation against the requirements

Commit:
refactor: move 'Reset' button to clear filter next to the filter selects

Commit:
feat: add sort options select component

This component should allow the user to select an option by which the paper are sorted. Currently, papers can be sorted by their id, title, year and decision; ascending and descending respectively.

Commit:
feat: add option to sort the project papers inside the stages on the papers overview page

Commit:
test: add unit test for project paper sorter function

Commit:
test: add integration test for 'SortOptionsSelect' component

Commit:
test: add e2e test for sorting papers on the project papers overview page

Commit:
feat: persist selected sort option for the project papers in the URL

Commit:
test: add unit and e2e tests for persisting the sort option in the URL

luca-schlecker commented on src/lib/components/composites/select/SortOptionsSelect.svelte:
```diff
@@ -0,0 +1,60 @@
+        ArrowDownZA,
+        ArrowUp,
+        ArrowUp10,
+        ArrowUpAZ,
```
The current icon selection is a bit confusing imo.
Sometimes the arrow changes direction, sometimes the decoration beside it.

What do you think about having it always pointing downwards and just swapping the order of the decorations?

| Id: low to high | Id: high to low | Title: low to high | Title: high to low |
| :---: | :---: | :---: | :---: |
| <img width="40" alt="image" src="https://github.com/user-attachments/assets/f1f17415-cb56-46fe-af8c-16c520ede9f1" /> | <img width="40" alt="image" src="https://github.com/user-attachments/assets/d885cee2-3c75-4114-b44b-715e2c7df6c3" /> |  <img width="40" alt="image" src="https://github.com/user-attachments/assets/fe59443b-bd9c-4c97-8089-b1c9151a6ed6" /> | <img width="40" alt="image" src="https://github.com/user-attachments/assets/005ea961-2e72-4003-99bf-b7538c8d63a0" /> |

Merseleo replied:
I also tested this solution and in my opinion it is more difficult to recognize the sorting direction when only the decorations change and not the arrow, but actually it doesn't matter to me, so I can change it, if you prefer this solution.

Merseleo replied:
Do have any further ideas for the decoration for arrow for indicating in which order the decisions are sorted?

luca-schlecker replied:
I'd prefer either my solution or where the order of the decorations is fixed and only the arrow changes.

Merseleo replied:
Ok, I always used an arrow pointing downwards, what do you think?

luca-schlecker replied:
Way better. :+1:

luca-schlecker commented on src/lib/components/composites/select/SortOptionsSelect.svelte:
```diff
@@ -0,0 +1,60 @@
+        {:else}
+            <ArrowDown10 class="size-4" />
+        {/if}
+    </Select.Trigger>
```
The trigger seems to have a slight border:
<img width="306" alt="image" src="https://github.com/user-attachments/assets/25973963-0b5f-4a36-9951-dca601246003" />

Merseleo replied:
You have a real eagle eye, I only realised that the second time I looked at it.

luca-schlecker commented on src/lib/model/sortCriteria.ts:
Files should be `kebab-case`.

Merseleo replied:
You are right

luca-schlecker commented on src/lib/model/sortCriteria.ts:
```diff
@@ -0,0 +1,71 @@
+        criterion: SortCriteria.YEAR,
+        direction: SortDirection.DESC,
+    },
+};
```
Is there a reason why this is in place to potentially prevent a user from sorting any criterion in the direction they'd like?

I feel like this should be inverted: a function taking a `SortOption` and returning the text currently used as a key.
The text input would then just list every possible combination of `SortCriteria` and `SortDirection`.

Let me know your thought's on this.

Merseleo replied:
How does this prevent a user from sorting any criterion? I think I do not quite understand what you want to tell me. Can you elaborate?

luca-schlecker replied:
You currently maintain a list of all possible combinations the user is allowed to sort by (in `ALLOWED_SORT_OPTIONS`). IMO this is not necessary as all possible combinations are valid.

Maybe the dropdown could look something like this?

![image](https://github.com/user-attachments/assets/68562b2e-c3a7-4aab-b57a-f909bda404bb)

This is only a rough sketch though, it should just demonstrate how a composition could look like. I am looking forward to your opinion on this.

Merseleo replied:
I did some further research and in my opinion this would be very unusual to have two buttons, but way more usual to have one select with all options.

Merseleo replied:
Furthermore I tried to implement your idea with the function and in my opinion this would complicate everything, especially when we want it type safe and we would loose flexibility when we do not want symmetric sort options.

luca-schlecker replied:
Ok. We'll leave it like this then. Asymmetric options seem rather weird, but I guess it would be more complex then. :+1:

luca-schlecker commented on src/lib/utils/sorters.ts:
```diff
@@ -0,0 +1,65 @@
+
+/**
+ * Compares two paper decisions based on a custom priority order:
+ * ACCEPTED \> IN_REVIEW \> DECLINED \> UNREVIEWED
```
I think this line is wrong. To me, it suggests that `ACCEPTED` > `UNREVIEWED`.
Yet, the current implementation orders `ACCEPTED` **<** `UNREVIEWED`, as `UNREVIEWED` has a higher numeric value.

Merseleo replied:
You are right, but if you look at the implementation of the sort function, lower values are sorted before higher values, so if you want to sort something ascending, things with a _higher_ priority should have _lower_ numeric values. However, if this is confusing in your opinion, if can assign ACCEPTED to 4, unspecified to 0 and invert the return statement: 
```typescript
function comparePaperDecisions(a: PaperDecision, b: PaperDecision): number {
    const priority: Record<PaperDecision, number> = {
        [PaperDecision.UNSPECIFIED]: 0,
        [PaperDecision.UNREVIEWED]: 1,
        [PaperDecision.DECLINED]: 2,
        [PaperDecision.IN_REVIEW]: 3,
        [PaperDecision.ACCEPTED]: 4,
    };

    // subtract the priority of b from the priority of a as if a has the higher priority the return value must be negative  for the sorting function
    return priority[b] - priority[a];
}
```

luca-schlecker replied:
No no, I agree with your decision to do it this way around with `ACCEPTED` having a value of `0`. Only the comment is misleading imo. Swap the `>` signs:
```
ACCEPTED < IN_REVIEW < DECLINED < UNREVIEWED
```

`ACCEPTED < IN_REVIEW` as $0 - 1 = -1 < 0$

Merseleo replied:
Ok, I adapted the comment. What do you think of it?

luca-schlecker replied:
Better now. :+1:

luca-schlecker commented on src/lib/utils/sorters.ts:
```diff
@@ -0,0 +1,65 @@
+                break;
+
+            case SortCriteria.PAPER_ID:
+                compareResult = a.localId.localeCompare(b.localId);
```
I know that `id` and `localId` currently are strings, however, I believe a user would expect `4` to come before `11` if sorted by id from low to high.

If I recall correctly, currently the plan for local ids is to be only of numerical nature.

I'd suggest trying to parse ids as integers. If both ids of two papers are integers, the lower one is ordered before. If both ids are strings, the current implementation can be used to order them. If, however, one is numerical and one not, the integer value should be ordered before the non-numerical.

Merseleo replied:
Good catch, I will implement your suggestion or even better I just found a function from @Slartibartfass2 that already compares two paper ids. Are you ok with the implementation of this function?

luca-schlecker replied:
I agree with the implementation. Failing to parse a numerical id will not sort it by lexicographical order, though.

I'll leave it up to you whether non-numerical ids are seen as equal or should be compared using lexicographical order. This function would implement the ordering I suggested, but you can modify it however you like:
```ts
function comparePaperId(a: string, b: string): number {
    const idA = parseInt(a.slice(1), 10);
    const idB = parseInt(b.slice(1), 10);

    const aNan = isNaN(idA);
    const bNan = isNaN(idB);

    if (!aNan && !bNan) return idA - idB;
    else if (aNan && bNan) return a.localCompare(b);
    else if (aNan && !bNan) return 1;
    else if (!aNan && bNan) return -1;
}
```

Merseleo replied:
Ok I though a while about your suggestion and did not found anything better.

luca-schlecker commented on tests/unit/utils/search-parameters.test.ts:
```diff
@@ -116,3 +115,40 @@ describe("Helper filters URL query parameters", () => {
+        expect(getSortOptionFromURL()).toBe("Title: Z to A");
+    });
+
+    test("When the query parameter for the sort option is not present, then undefined will be returned.", async () => {
```
I cannot see how "undefined" plays any role in this test?
This should probably state that the default is expected.

Merseleo replied:
Good catch, this is a leftover from the implementation phase. First, I returned undefined and checked somewhere else for undefined and handled the case, but corrected it to return a default value, if something went wrong. Thank you for the hint.

luca-schlecker reviewed (changes requested):
Great work! Works well, is documented and has tests.
There are a few minor questions and issues, though.
I am looking forward to hearing your opinion on them.

Commit:
test: correct test case description for test that query parameter for sort option is not present.

The description was not updated as the default behavior of the 'getSortOptionFromURL' function changed from returning undefined to returning a meaningful default value.

Commit:
refactor: remove the border from the select trigger and change the icons indicating the sort criteria and direction

Commit:
refactor: rename sortCriteria.ts into sort-criteria.ts

Commit:
fix: use 'comparePaperId' function to compare the local paper ids when sorting project papers

Using the 'localCompare' method does not work when the Ids are greater than 10, because then 10 comes before 9 if sorted ascending which is not correct.

Merseleo reviewed (commented):
Thank you for the feedback and good ideas. The review was very attentive and I didn't notice some mistakes on my part, very good 👍

Commit:
docs: clarify comment for 'comparePaperDecisions' function

Commit:
refactor: adapt 'comparePaperId' function to not expecting a '#' in front of the ID and handle the case, when at least one ID is not numeric

Commit:
feat: add custom icons for the arrows indicating the sort direction for the criterion 'Decision'

Merseleo reviewed (commented):
I still have struggles to understand your ideas correctly, because in my opinion they would introduce further complexity and worsen user experience.

luca-schlecker reviewed (dismissed):
Everything seems to be addressed. Great work :+1:.
After a rebase, this can be merged. :rocket:

Merseleo commented:
I also asked Alex for his opinion on the design: He said that a select component in which every option is described in detail is probably easier to understand, especially with the _Decision_ option.

Commit:
refactor: add exhaustive check to 'sortProjectPaper' method when checking all sort criteria and prevent variable shadowing

Commit:
test: fix flaky test to check whether the paper view was opened after double clicking a paper

luca-schlecker reviewed (approved):
Great work, LGTM :rocket:
